// gzip-brunch/lib/index.js
// Generated by CoffeeScript 1.12.7
var Gzip, fs, sysPath, zlib,
  bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

fs = require('fs');

zlib = require('zlib');

sysPath = require('path');

module.exports = Gzip = (function() {
  Gzip.prototype.brunchPlugin = true;

  function Gzip(config) {
    var ref, ref1, ref2, ref3, ref4, ref5, ref6;
    this.config = config;
    this._joinToPublic = bind(this._joinToPublic, this);
    this.options = (ref = (ref1 = this.config) != null ? (ref2 = ref1.plugins) != null ? ref2.gzip : void 0 : void 0) != null ? ref : {};
    this.targets = [
      {
        path: this.config.paths["public"],
        ext: /\.html$/
      }, {
        path: this._joinToPublic(((ref3 = this.options.paths) != null ? ref3.javascript : void 0) || ''),
        ext: /\.js$/
      }, {
        path: this._joinToPublic(((ref4 = this.options.paths) != null ? ref4.stylesheet : void 0) || ''),
        ext: /\.css$/
      }, {
        path: this._joinToPublic(((ref5 = this.options.paths) != null ? ref5.image : void 0) || 'images'),
        ext: /\.svg$/
      }, {
        path: this._joinToPublic(((ref6 = this.options.paths) != null ? ref6.root : void 0) || ''),
        ext: /\.ico$/
      }
    ];
  }

  Gzip.prototype.onCompile = function(generatedFiles) {
    var fileList, i, len, ref, results, target;
    if (!this.config.optimize) {
      return;
    }
    ref = this.targets;
    results = [];
    for (i = 0, len = ref.length; i < len; i++) {
      target = ref[i];
      if (!fs.existsSync(target.path)) {
        break;
      }
      fileList = fs.readdirSync(target.path);
      results.push(fileList.forEach((function(_this) {
        return function(file) {
          if (file.match(target.ext)) {
            return _this._compress(target.path, file);
          }
        };
      })(this)));
    }
    return results;
  };

  Gzip.prototype._compress = function(path, file) {
    var gzip, input, input_path, output, output_path;
    gzip = zlib.createGzip({
      level: zlib.Z_BEST_COMPRESSION
    });
    input_path = path + "/" + file;
    output_path = path + "/" + file + ".gz";
    input = fs.createReadStream(input_path);
    output = fs.createWriteStream(output_path);
    input.pipe(gzip).pipe(output);
    if (!!this.options.removeOriginalFiles && fs.existsSync(input_path)) {
      fs.unlinkSync(input_path);
    }
    if (!!this.options.renameGzipFilesToOriginalFiles && fs.existsSync(input_path)) {
      return fs.renameSync(output_path, input_path);
    }
  };

  Gzip.prototype._joinToPublic = function(path) {
    return sysPath.join(this.config.paths["public"], path);
  };

  return Gzip;

})();
